[aoti] Add initial custom op support

angelayi opened the pull request:
Re-land of https://github.com/pytorch/pytorch/pull/125242

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

swolchok commented on BUILD.bazel:
```diff
@@ -580,6 +580,7 @@ cc_library(
         "@eigen",
         "@fbgemm//:fbgemm_src_headers",
         "@fmt",
+        "@nlohmann",
```
adding a new third-party dep could be a separate PR

swolchok commented on test/inductor/test_aot_inductor.py:
```diff
@@ -2910,7 +2925,21 @@ def forward(self, x, y):
 
 
-class AOTInductorTestABICompatibleCpu(TestCase):
+class CustomTestCase(TestCase):
```
nit: all code we write is custom. How about CustomOpTestCase?

swolchok commented on test/inductor/test_aot_inductor_utils.py:
```diff
@@ -49,6 +49,15 @@ def compile(
+            from deeplearning.aot_inductor.extern_node_thrift_serializer import (
+                thrift_serializer,
+            )
+            if options is None:
+                options = {"extern_node_serializer": thrift_serializer}
+            else:
+                options["extern_node_serializer"] = thrift_serializer
```
nit: you can avoid repeating yourself like so

```
if options is None:
    options = {}
options["extern_node_serializer"] = thrift_serializer
```

(yes it's slightly slower, but that battle was already lost when we started writing Python)

swolchok commented on torch/csrc/inductor/aoti_runner/model_container_runner.cpp:
```diff
@@ -46,6 +47,18 @@ AOTIModelContainerRunner::AOTIModelContainerRunner(
 
+  // Hack to find the json file name from the model so file
+  size_t lastindex = model_so_path.find_last_of(".");
+  const std::string& json_filename =
```
this only happens to work because of an arcane C++ quirk (temporary lifetime extension -- https://abseil.io/tips/107); `std::string json_filename = ...` would be clearer

swolchok commented on torch/csrc/inductor/aoti_torch/oss_proxy_executor.cpp:
```diff
@@ -0,0 +1,266 @@
+void OSSProxyExecutor::prefill_stack_with_static_arguments(
+    int index,
+    at::TypePtr schema_arg_type,
+    json serialized_arg,
```
I've never used nlohmann before, but surely the `json` type isn't cheap to copy and therefore should be passed by const reference?

swolchok commented on torch/csrc/inductor/aoti_torch/oss_proxy_executor.cpp:
```diff
@@ -0,0 +1,266 @@
+// Populates op_kernel.outputs_
+void OSSProxyExecutor::get_output_info_from_serialized(
+    const std::vector<c10::Argument>& schema_returns,
+    json serialized_node,
```
ditto pass by reference

swolchok commented on torch/csrc/inductor/aoti_torch/oss_proxy_executor.cpp:
```diff
@@ -0,0 +1,266 @@
+
+    std::string opName;
+    std::string overloadName;
+    size_t pos = target.find(".");
```
nit: use a character literal, not a string literal (`'.'`)

swolchok commented on torch/csrc/inductor/aoti_torch/oss_proxy_executor.cpp:
```diff
@@ -0,0 +1,266 @@
+      overloadName = "";
+    } else {
+      // There should be no more periods
+      size_t pos2 = target.find(".", pos);
```
ditto character literal

swolchok commented on torch/csrc/inductor/aoti_torch/oss_proxy_executor.cpp:
```diff
@@ -0,0 +1,266 @@
+
+  int tensor_id = 0;
+  int int_id = 0;
+  for (size_t i = 0; i < dynamic_args.size(); i++) {
```
nit: you don't need the value of `i`, so why not use a range-for loop?

swolchok commented on torch/csrc/inductor/aoti_torch/oss_proxy_executor.h:
```diff
@@ -0,0 +1,99 @@
+      : arg_index(arg_index),
+        arg_type(arg_type),
+        length(length),
+        serialized_arg_val(serialized_arg_val) {}
```
missing std::move

swolchok commented on torch/csrc/inductor/aoti_torch/oss_proxy_executor.cpp:
```diff
@@ -0,0 +1,266 @@
+      TORCH_CHECK(serialized_arg_type == "as_tensor");
+      stack.emplace_back();
+      dynamic_args.emplace_back(
+          index, DynamicArgType::TensorType, 1, serialized_arg_val);
```
missing std::move(serialized_arg_val)

more importantly, IIUC we are going to parse JSON once at startup and then use it in a potentially-performance-critical path, potentially very many times. Shouldn't we extract the data into plain old structs to save memory and hash-table-lookups later?

angelayi replied:
Ah yes that makes sense! I can address this in a follow up!

swolchok reviewed (commented):
not well-versed in inductor; just some general comments + higher-level issue of JSON parsing

malfet commented on torch/csrc/inductor/aoti_torch/oss_proxy_executor.h:
```diff
@@ -0,0 +1,99 @@
+
+using json = nlohmann::json;
+
+namespace torch {
+namespace aot_inductor {
```
Nit
```suggestion
namespace torch::aot_inductor {
```

malfet commented on torch/csrc/inductor/aoti_torch/oss_proxy_executor.h:
```diff
@@ -0,0 +1,99 @@
+namespace torch {
+namespace aot_inductor {
+
+enum DynamicArgType : int {
```
Not to leak it in a global namespace
```suggestion
enum class DynamicArgType : int {
```

malfet commented on torch/csrc/inductor/aoti_torch/oss_proxy_executor.h:
```diff
@@ -0,0 +1,99 @@
+  std::unique_ptr<c10::Device> device_;
+};
+
+} // namespace aot_inductor
+} // namespace torch
```
```suggestion
} // namespace torch::aot_inductor
```

malfet commented on torch/csrc/inductor/aoti_runner/model_container_runner.cpp:
```diff
@@ -2,8 +2,17 @@
+#include <torch/csrc/inductor/aoti_torch/oss_proxy_executor.h>
 #include <torch/csrc/inductor/aoti_torch/tensor_converter.h>
 
+#if __has_include("filesystem")
```
```suggestion
// TODO: Investigate why this is necessary, but fixes build problems in FRL
#if __has_include("filesystem")
```

malfet commented on torch/csrc/inductor/aoti_torch/oss_proxy_executor.cpp:
```diff
@@ -0,0 +1,266 @@
+}
+} // namespace
+
+namespace torch {
+namespace aot_inductor {
```
```suggestion
namespace torch::aot_inductor {
```

malfet commented on torch/csrc/inductor/aoti_torch/oss_proxy_executor.cpp:
```diff
@@ -0,0 +1,266 @@
+      num_tensors);
+}
+
+} // namespace aot_inductor
+} // namespace torch
```
```suggestion
} // namespace torch::aot_inductor
```

malfet commented on torch/csrc/inductor/aoti_torch/oss_proxy_executor.h:
```diff
@@ -0,0 +1,99 @@
+#include <torch/csrc/inductor/aoti_torch/c/shim.h>
+#include <torch/csrc/inductor/aoti_torch/proxy_executor.h>
+
+using json = nlohmann::json;
```
Not using in headers please
```suggestion
```

malfet commented on torch/csrc/inductor/aoti_torch/oss_proxy_executor.h:
```diff
@@ -0,0 +1,99 @@
+      int arg_index,
+      DynamicArgType arg_type,
+      int length,
+      json serialized_arg_val)
```
```suggestion
      nlohmann::json serialized_arg_val)
```

malfet commented on torch/csrc/inductor/aoti_torch/oss_proxy_executor.cpp:
```diff
@@ -0,0 +1,266 @@
+    } else {
+      TORCH_CHECK(
+          false,
+          "NYI: Unsupported return type for schema: ",
```
What NYI stands for?

angelayi replied:
not yet implemented!

malfet commented on torch/csrc/inductor/aoti_torch/oss_proxy_executor.h:
```diff
@@ -0,0 +1,99 @@
+  int arg_index;
+  DynamicArgType arg_type;
+  int length;
+  json serialized_arg_val;
```
```suggestion
  nlohmann::json serialized_arg_val;
```

malfet commented on torch/csrc/inductor/aoti_torch/oss_proxy_executor.h:
```diff
@@ -0,0 +1,99 @@
+  std::vector<DynamicArg> outputs_;
+  std::vector<c10::IValue> stack_;
+
+  int num_output_tensors() const {
```
Why signed value?

angelayi replied:
I just followed [these signed values](https://github.com/pytorch/pytorch/blob/f093cd40864d5c8d4ceecb72fc9828a9fcb94a4e/torch/csrc/inductor/aoti_torch/proxy_executor.h#L15-L20) 😅 not sure if those should also be changed to size_t

malfet commented on torch/csrc/inductor/aoti_torch/oss_proxy_executor.h:
```diff
@@ -0,0 +1,99 @@
+
+  void get_input_info_from_serialized(
+      const std::vector<c10::Argument>& schema_args,
+      const json& serialized_node,
```
```suggestion
      const nlohmann::json& serialized_node,
```

malfet commented on torch/csrc/inductor/aoti_torch/oss_proxy_executor.h:
```diff
@@ -0,0 +1,99 @@
+
+  void get_output_info_from_serialized(
+      const std::vector<c10::Argument>& schema_returns,
+      const json& serialized_node,
```
```suggestion
      const nlohmann::json& serialized_node,
```

angelayi committed:
[aoti] Add initial custom op support (#127034)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/127034

Reviewed By: malfet

Differential Revision: D57753395

Pulled By: angelayi

angelayi commented:
@pytorchbot merge
